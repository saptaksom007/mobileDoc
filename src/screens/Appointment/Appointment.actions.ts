export const ActionTypes = {
  APPOINTMENT_REQUESTED: 'APPOINTMENT_REQUESTED',
  APPOINTMENT_SUCCEEDED: 'APPOINTMENT_SUCCEEDED',
  APPOINTMENT_FAILED: 'APPOINTMENT_FAILED'
}

export default class Actions {
  static appointmentRequested(id: number) {
    return {
      type: ActionTypes.APPOINTMENT_REQUESTED,
      id
    }
  }
  static appointmentSucceeded(payload: any) {
    return {
      type: ActionTypes.APPOINTMENT_SUCCEEDED,
      payload
    }
  }
  static appointmentFailed(error: Error) {
    return {
      type: ActionTypes.APPOINTMENT_FAILED,
      payload: {
        error
      }
    }
  }
}
