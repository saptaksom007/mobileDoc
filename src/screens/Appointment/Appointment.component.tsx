import React, { Component } from 'react'
import { View, InteractionManager, ScrollView } from 'react-native'
import { connect } from 'react-redux'

import { compose } from 'recompose'
import { propOr } from 'ramda'
import Clinic from 'components/Clinic/Clinic.component'
import { Avatar } from 'components/Avatar/Avatar.component'
import MessageMyDoctor from 'components/MessageMyDoctor/MessageMyDoctor.component'
import ExportAppointmentLink from 'components/ExportAppointmentLink/ExportAppointmentLink.component'
import {
  Paragraph,
  Bold,
  Medium,
} from 'components/CustomText/CustomText.component'
import { fontFamily } from 'components/CustomText/CustomText.styles'
import { Card, ListItem } from 'react-native-elements'
import withSubtitle from 'screens/enhancers/withSubtitle'
import withoutKeyboard from 'screens/enhancers/withoutKeyboard'
import { withLoader } from 'screens/enhancers/withLoader'
import i18n from 'ex-react-native-i18n'
import { findPatientSelf } from 'common-docdok/lib/domain/healthrelation/selectors/findPatient'
import { findConversationForHealthSubjectUuid } from 'common-docdok/lib/domain/messaging/selectors/findConversation'
import {
  momentWithLocales as moment,
  getCalendarOutputFormatWithToday,
} from 'utilities/date'

import NotificationsActions from 'notifications/actions'
import toString from 'common-docdok/lib/utils/toString'

import Actions from './Appointment.actions'

import styles from './Appointment.styles'
import { NavigationScreenProp } from 'react-navigation'
import { Action } from 'redux'
import { ConversationDtoType } from 'common-docdok/lib/types'

function getSafeFromAppointment(key: string, appointment: any): string {
  return propOr(i18n.t('global.undefined'), key, appointment)
}

function getAppointmentTime(appointment: any) {
  return moment(appointment.appointmentTime || '2017-06-19').locale(
    i18n.getFallbackLocale(),
  )
}

interface Props {
  dispatch(action: Action): void
  navigation: NavigationScreenProp<{
    params: { appointmentId: number }
  }>
  appointment: any & {
    physician: string
    physicianAvatarPicture: string
    patientName: string
    message?: string
  }
  name: string
  patientSelf: any
  conversationId: number
}

export class Appointment extends Component<Props> {
  componentDidMount() {
    InteractionManager.runAfterInteractions(() => {
      const {
        navigation: {
          state: { params },
        },
        dispatch,
      } = this.props
      dispatch(Actions.appointmentRequested(params && params.appointmentId))
      dispatch(NotificationsActions.removeNotificationRequested('APPOINTMENT'))
    })
  }

  renderDoctorAndSchedule() {
    const { appointment } = this.props
    const appointmentTime = getAppointmentTime(appointment)
    const physician = getSafeFromAppointment('physician', appointment)
    return (
      <Medium style={styles.message}>
        <Bold>{i18n.t('appointment.card.doctor')}:</Bold> {physician}
        {'\n'}
        <Bold>{i18n.t('appointment.card.date')}:</Bold>{' '}
        {appointmentTime.format('L')}
        {'\n'}
        <Bold>{i18n.t('appointment.card.time')}:</Bold>{' '}
        {appointmentTime.format('LT')}
      </Medium>
    )
  }

  renderMessage() {
    const { appointment } = this.props
    const message = getSafeFromAppointment('message', appointment)
    return (
      <Medium style={styles.message}>
        <Bold>{i18n.t('appointment.card.message')}:</Bold> {message}
      </Medium>
    )
  }

  renderClinic() {
    const { appointment } = this.props
    const clinic = getSafeFromAppointment('clinicName', appointment)
    const street = getSafeFromAppointment('clinicStreet', appointment)
    const location = getSafeFromAppointment('clinicLocation', appointment)
    const phoneNumber = getSafeFromAppointment('clinicContact', appointment)
    return (
      <Clinic
        name={clinic}
        contact={phoneNumber}
        street={street}
        location={location}
      />
    )
  }

  render() {
    const { appointment, name, patientSelf, conversationId } = this.props

    if (!appointment) {
      return <View />
    }

    const avatar = (
      <Avatar
        name={appointment.physician}
        url={appointment.physicianAvatarPicture}
      />
    )

    const appointmentTime = getAppointmentTime(appointment)
    return (
      <ScrollView contentContainerStyle={styles.contentContainerStyle}>
        <View style={styles.messageContainer}>
          <Paragraph style={styles.message}>
            <Paragraph>
              {i18n.t('appointment.salutation', {
                name,
                default: 'Hello,',
              })}
            </Paragraph>
            {'\n'}
            {'\n'}
            {patientSelf && (
              <Paragraph>
                {i18n.t('appointment.message', {
                  physician: appointment.physician,
                  hour: appointmentTime.format('LT'),
                  day: appointmentTime.format('dddd'),
                  date: appointmentTime.format('Do'),
                  month: appointmentTime.format('MMMM'),
                  calendar: appointmentTime.calendar(
                    undefined,
                    getCalendarOutputFormatWithToday(),
                  ),
                })}
              </Paragraph>
            )}
            {!patientSelf && (
              <Paragraph>
                {i18n.t('appointment.messageProxy', {
                  physician: appointment.physician,
                  name: appointment.patientName,
                  hour: appointmentTime.format('LT'),
                  day: appointmentTime.format('dddd'),
                  date: appointmentTime.format('Do'),
                  month: appointmentTime.format('MMMM'),
                  calendar: appointmentTime.calendar(
                    undefined,
                    getCalendarOutputFormatWithToday(),
                  ),
                })}
              </Paragraph>
            )}
          </Paragraph>
        </View>
        <Card
          containerStyle={styles.cardContainerStyle}
          title={i18n.t('appointment.card.title')}
          titleStyle={styles.cardTitle}
          fontFamily={fontFamily}
        >
          <ListItem
            avatar={avatar}
            roundAvatar
            titleNumberOfLines={3}
            title={this.renderDoctorAndSchedule()}
            titleContainerStyle={styles.cardFirstItemTitle}
            hideChevron
            containerStyle={styles.cardLastItem}
          />

          <ListItem
            title={this.renderClinic()}
            titleContainerStyle={styles.cardFirstItemTitle}
            hideChevron
            containerStyle={[
              styles.cardLastItem,
              !appointment.message && styles.cardLastItemNoBorder,
            ]}
          />

          {appointment.message && (
            <ListItem
              title={this.renderMessage()}
              titleContainerStyle={styles.cardFirstItemTitle}
              hideChevron
              containerStyle={[
                styles.cardLastItem,
                styles.cardLastItemNoBorder,
              ]}
            />
          )}
        </Card>
        <View style={styles.bottomLinks}>
          <MessageMyDoctor
            containerStyle={styles.messageMyDoctor}
            forceConversationId={conversationId}
          />
          <ExportAppointmentLink appointment={appointment} />
        </View>
      </ScrollView>
    )
  }
}

function select({
  appointment: { data },
  profile,
  healthrelation: { patients },
  messaging: { conversations },
}: any): any {
  const patientSelf = patients.find(findPatientSelf)
  const name = (profile && ` ${toString.person(profile)}`) || ''
  const appointment = data

  const conversationsArray = Object.values(
    conversations,
  ) as ConversationDtoType[]
  const conversation: any =
    appointment &&
    conversationsArray.find(
      findConversationForHealthSubjectUuid(appointment.healthSubjectUuid),
    )

  return {
    appointment,
    name,
    conversationId: conversation && conversation.id,
    patientSelf:
      patientSelf &&
      appointment &&
      appointment.healthSubjectUuid === patientSelf.uuid,
  }
}

export default compose<Props, any>(
  withoutKeyboard,
  connect(select),
  withLoader,
  withSubtitle('appointment.subtitle'),
)(Appointment)
