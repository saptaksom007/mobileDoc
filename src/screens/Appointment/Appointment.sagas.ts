import { call, put, takeEvery } from 'redux-saga/effects'
import { Actions as NavAction } from 'navigation/SagaNavigation'
import i18n from 'ex-react-native-i18n'

import Actions from './Appointment.actions'
import { fetchAppointmentByIdAsync } from 'api/appointments'

function* fetchAppointment({ id }: any) {
  try {
    const appointment = yield call(fetchAppointmentByIdAsync, { id })
    yield put(Actions.appointmentSucceeded(appointment))
  } catch (err) {
    yield put(Actions.appointmentFailed(err))
    yield put(
      NavAction.showLocalError(i18n.t('appointment.errors.fetchAppointment')),
    )
  }
}

export default [takeEvery('APPOINTMENT_REQUESTED', fetchAppointment)]
