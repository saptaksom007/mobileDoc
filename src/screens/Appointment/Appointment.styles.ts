import { StyleSheet } from 'react-native'
import { Layout } from 'constants/Layout'
import { Color } from 'constants/Color'

export default StyleSheet.create({
  contentContainerStyle: {
    height: Layout.window.height
  },
  container: {
    flex: 1,
    justifyContent: 'flex-start',
    backgroundColor: '#fff'
  },
  messageContainer: {
    backgroundColor: '#fff',
    marginTop: 25,
    margin: 15,
    justifyContent: 'flex-start',
    alignItems: 'flex-start'
  },
  message: {
    textAlign: 'left'
  },
  cardContainerStyle: {
    marginTop: 10
  },
  cardTitle: {
    color: Color.lightText
  },
  cardLastItem: {
    paddingTop: 5,
    paddingBottom: 5
  },
  cardLastItemNoBorder: {
    borderBottomWidth: 0
  },
  cardFirstItemTitle: {
    marginHorizontal: 10
  },
  messageMyDoctor: {
    marginTop: 0,
    width: undefined,
    height: undefined
  },
  bottomLinks: {
    flexDirection: 'row',
    width: Layout.window.width,
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 30
  }
})
