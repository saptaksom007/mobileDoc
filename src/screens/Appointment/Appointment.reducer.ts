import { ActionTypes } from './Appointment.actions'

const { APPOINTMENT_REQUESTED, APPOINTMENT_SUCCEEDED } = ActionTypes

interface State {
  error?: Error
  data?: any
  id?: number
}

type Action = {
  type: keyof typeof ActionTypes
  payload: any
  id: number
}

const initialState: State = {
  error: undefined,
  data: undefined,
  id: undefined
}

export default function appointment(
  state: State = initialState,
  action: Action
) {
  switch (action.type) {
    case APPOINTMENT_REQUESTED: {
      const { id } = action
      return { ...state, error: undefined, id }
    }

    case APPOINTMENT_SUCCEEDED: {
      const { payload } = action
      return {
        ...state,
        error: undefined,
        data: payload
      }
    }

    default:
      return state
  }
}
