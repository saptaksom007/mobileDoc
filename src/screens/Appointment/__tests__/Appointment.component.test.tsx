import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { Appointment } from '../Appointment.component'
import { NavigationScreenProp } from 'react-navigation'

jest.mock(
  'components/MessageMyDoctor/MessageMyDoctor.component',
  () => 'MessageMyDoctor'
)
jest.mock(
  'components/ExportAppointmentLink/ExportAppointmentLink.component',
  () => 'ExportAppointmentLink'
)

it('<Appointment/> render correctly', () => {
  const tree = renderer
    .create(
      <Appointment
        navigation={
          { state: { params: { appointmentId: 0 } } } as NavigationScreenProp<
            any
          >
        }
        dispatch={() => {}}
        appointment={{}}
        name=""
        patientSelf={{}}
        conversationId={0}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
