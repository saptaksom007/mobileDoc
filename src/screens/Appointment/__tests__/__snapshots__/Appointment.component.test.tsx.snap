// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Appointment/> render correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "height": 1334,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#fff",
          "justifyContent": "flex-start",
          "margin": 15,
          "marginTop": 25,
        }
      }
    >
      <Text
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontFamily": "TitilliumWeb-Regular",
              "fontSize": 17.5,
              "lineHeight": 22,
              "overflow": "visible",
            },
            Object {
              "color": "rgba(27, 106, 158, 0.85)",
              "fontSize": 17.5,
              "lineHeight": 21,
            },
            Object {
              "textAlign": "left",
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 17.5,
                "lineHeight": 22,
                "overflow": "visible",
              },
              Object {
                "color": "rgba(27, 106, 158, 0.85)",
                "fontSize": 17.5,
                "lineHeight": 21,
              },
              undefined,
            ]
          }
        >
          appointment.salutation
        </Text>
        

        

        <Text
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 17.5,
                "lineHeight": 22,
                "overflow": "visible",
              },
              Object {
                "color": "rgba(27, 106, 158, 0.85)",
                "fontSize": 17.5,
                "lineHeight": 21,
              },
              undefined,
            ]
          }
        >
          appointment.message
        </Text>
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 30,
          "width": 750,
        }
      }
    >
      <MessageMyDoctor
        containerStyle={
          Object {
            "height": undefined,
            "marginTop": 0,
            "width": undefined,
          }
        }
        forceConversationId={0}
      />
      <ExportAppointmentLink
        appointment={Object {}}
      />
    </View>
  </View>
</RCTScrollView>
`;
